woocommerce/woocommerce-android
# Feature: Sign In Milestone 1

This PR fixes #1091 by implementing the latest designs for Sign in milestone 1. This feature PR includes the following approved PRs:
- [Part 1: Sign In Jetpack basic check and error messaging](https://github.com/woocommerce/woocommerce-android/pull/1121)
- [Part 2: What is Jetpack Dialog](https://github.com/woocommerce/woocommerce-android/pull/1125)
- [Part 3: Sign In post authentication validate url and auto-login](https://github.com/woocommerce/woocommerce-android/pull/1135)
- [Part 4: Sign in Milestone 1 Tracks Events](https://github.com/woocommerce/woocommerce-android/pull/1137)
- [Part 5: Signin M1 - add jetpack support documentation](https://github.com/woocommerce/woocommerce-android/pull/1147)

### New build config variable for debug builds
- `wp.debug.wpcom_website_url`: This will pre-populate the site address for debug builds

# Major Design Flows
Aside from the normal url scrubbing, below are the major flows associated with this project.
- **WordPress Required** -> Site at URL does not have WordPress installed
- **Jetpack Required** -> Site at URL does not have Jetpack
- **WooCommerce Required** -> Set at URL does not have WooCommerce installed
- **Account Mismatch** -> Site at URL does not belong to the authenticated account
- **Auto-Login Success** -> Site at URL is a Jetpack-connected WordPress site with the WooCommerce plugin installed and activated **AND** it's connected to the authenticated account.

## WordPress Required
<img src="https://user-images.githubusercontent.com/5810477/59474376-c30d8500-8e03-11e9-8850-65df24f08b01.png" width="300"/>

## Jetpack Required
<img src="https://user-images.githubusercontent.com/5810477/59474249-54302c00-8e03-11e9-969b-0b23f203a993.gif" width="300"/>

<img src="https://user-images.githubusercontent.com/5810477/59474306-74f88180-8e03-11e9-8564-debaf1f97e04.gif" width="300"/>

## WooCommerce Required
<img src="https://user-images.githubusercontent.com/5810477/59474878-d0c40a00-8e05-11e9-9f82-468f6d491083.gif" width="300"/>

## Account Mismatch
<img src="https://user-images.githubusercontent.com/5810477/59474968-4203bd00-8e06-11e9-8067-8db004c3968c.gif" width="300"/>


## Auto-Login Success!
<img src="https://user-images.githubusercontent.com/5810477/59474988-58aa1400-8e06-11e9-8f8c-472e1394ae5e.gif" width="300"/>


# Design Deviations

### Check for WooCommerce
The designs required checking for WordPress, Jetpack, and Woo when the user submits the url for login, but this would've required a huge amount of code so for this iteration, so we decided to keep it simple since this new login flow is a test and we may or may not keep it. So the temporary workaround is to use an existing FluxC endpoint: `FetchConnectedSiteInfo`. This will tell us if the site exists, has WP, and has Jetpack connected, **but with some caveats to be aware of:**
- Sites requiring HTTP Auth will not work
- Self-signed certificates will not work
If in the future we decide to keep this flow, a more robust solution will need to be developed to ensure all scenarios are accounted for.

Checking if the site has WooCommerce installed is not available until _after_ the account is authenticated so that part of the flow has been moved to post-authentication.

### Offline error messaging
The original designs called for an error message to be displayed below the text input if the device was not connected, but since WCAndroid uses a shared library I wanted to make sure it was okay with the folks over at WPAndroid. At the end it was decided to leave it as-is. So if the device is not connected, then the user will just see a toast message.

# Other features
- landscape versions of each view have also been implemented where appropriate.
- **View connected stores** button is only visible if the authenticated account has connected woo stores.
<img src="https://user-images.githubusercontent.com/5810477/59475038-a1fa6380-8e06-11e9-883b-2b484a184a38.png" width="300"/>

# To Test
Each of the PRs that make this feature were independently tested, but here's an overview of things that should be verified for alpha/beta:
- Each of the main stated flows using various login options (magic link, google (if release build), password). 
- Landscape modes
- Clicking through each of the help dialog options
- Switching sites (since `SitePickerActivity` was so heavily modified)
- Invalid URL (empty, spaces...etc)
- No active site at url
- Tracks events

Update release notes:
